rfd/0131-adminitrative-actions-mfa.md

```diff
@@ -0,0 +1,532 @@
+solve the returned challenge, and attach it to the admin action request.
+
+If a user has no MFA device registered, `CreateAuthenticateChallenge` will fail.
+In this case, the client will make the request without the MFA challenge response,
+in case we are handling a special case (e.g. Built in `Admin` role, bot user
+for automated use cases).
+
+Additionally, if the client doesn't whether an action requires MFA or not, it
+will first attempt without it. If the server responds with `ErrAdminActionMFARequired`,
+the client will attempt to retry the request with MFA.
+
+### Proto Specification
+
+Each Admin action API request will need to include the `MFAAuthenticateResponse`
+field in its proto specification.
+
+Clients will automatically check for the `MFAAuthenticateResponse` in order to
+determine whether it must prompt for MFA. This will be done seamlessly within
+a `grpc.UnaryInterceptor`.
+
+Likewise, the Server will automatically check for an `MFAAuthenticateResponse` in
+using a `grpc.UnaryInterceptor`. If found, this will be passed down the chain
+through the `AuthContext` (like user) where it can be verified by `ServerWithRoles`
+if required.
+
+#### HTTP Endpoints
+
+Some requests which we would like to make admin actions are still used through
+the HTTP API rather than the gRPC API, such as `POST UpsertUser`. Part of this
+work will require converting these requests to gRPC, which will also move us
+closer to [fully deprecating the HTTP API](https://github.com/gravitational/teleport/issues/6394).
+
+#### Non-unique gRPC Requests
+
+Many any of our existing gRPC API requests do not use unique request or response
+messages, despite it being a best practice. As a result, it is not trivial to add
+the `MFAAuthenticateResponse` field to some requests.
+
+For example:
+
+```proto
+rpc UpdateUser(types.UserV2) returns (google.protobuf.Empty);
+```
+
+For these cases, we will need to create replacement requests:
+
+```proto
+rpc UpdateUser(types.UserV2) returns (google.protobuf.Empty);
+rpc UpdateUserV2(UpdateUserV2Request) returns (UpdateUserV2Response);
+```
+
+For many of these cases above, it will make more sense to add or move the rpc
+into a new service. This will allow us to:
+
+- Maintain preferred naming conventions, avoiding `V2` suffixes
+- Reduce reliance on gogoproto which has got to go
+- Reduce size of legacy proto packages which are incompatible with buf
+
+This will be handled on a case by case basis, but user service looks like a
+likely candidate:
+
+```proto
+service UserService {
+  rpc GetUser(GetUserRequest) returns (types.UserV2);
+  rpc GetUsers(GetUsersRequest) returns (stream types.UserV2);
+  // We can replace CreateUser and UpdateUser with UpsertUser, as is the latest convention of our API.
+  rpc UpsertUser(UpsertUserRequest) returns (UpsertUserResponse);
+  rpc DeleteUser(DeleteUserRequest) returns (google.protobuf.Empty);
+  ...
+}
+```
+
```

Comment:
Can you also add section about product metrics and observability? I.e. how are we going to be able to tell whether people are using this feature? Can we emit a Posthog event in the Cloud every time someone tries to perform administrative action and has to tap the key?